File: uas-ppb-c030324023-backend/app/Models/Application.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

#[Fillable([
    'account_id', 'program_id', 'full_name', 'birth_place', 'birth_date', 'gender',
    'address', 'phone', 'school_origin', 'father_name', 'father_job', 'mother_name',
    'mother_job', 'parents_income', 'photo_path', 'status', 'edits_used', 'last_submitted_at',
])]
class Application extends Model
{
    protected function casts(): array
    {
        return [
            'birth_date' => 'date',
            'last_submitted_at' => 'datetime',
        ];
    }

    public function account(): BelongsTo
    {
        return $this->belongsTo(Account::class);
    }

    public function program(): BelongsTo
    {
        return $this->belongsTo(Program::class);
    }

    public function editableUntil(): \Carbon\Carbon
    {
        return $this->last_submitted_at->copy()->addMinutes(10);
    }

    public function editsRemaining(): int
    {
        return max(0, 3 - $this->edits_used);
    }

    // ponytail: verdict (status != submitted) takes priority over budget/time,
    // matching "admin verdict locks regardless of time/edit count remaining".
    public function lockReason(): ?string
    {
        if ($this->status !== 'submitted') {
            return 'APPLICATION_LOCKED';
        }
        if ($this->editsRemaining() === 0) {
            return 'EDITS_EXHAUSTED';
        }
        if (now()->greaterThan($this->editableUntil())) {
            return 'APPLICATION_LOCKED';
        }

        return null;
    }

    public function isLocked(): bool
    {
        return $this->lockReason() !== null;
    }
}

File: uas-ppb-c030324023-mobile/lib/models/application.dart
class Application {
  final int id;
  final int programId;
  final String fullName;
  final String birthPlace;
  final String birthDate;
  final String gender;
  final String address;
  final String phone;
  final String schoolOrigin;
  final String fatherName;
  final String fatherJob;
  final String motherName;
  final String motherJob;
  final String parentsIncome;
  final String photoUrl;
  final String status;
  final int editsRemaining;
  final bool locked;
  final String? editableUntil;

  Application({
    required this.id,
    required this.programId,
    required this.fullName,
    required this.birthPlace,
    required this.birthDate,
    required this.gender,
    required this.address,
    required this.phone,
    required this.schoolOrigin,
    required this.fatherName,
    required this.fatherJob,
    required this.motherName,
    required this.motherJob,
    required this.parentsIncome,
    required this.photoUrl,
    required this.status,
    required this.editsRemaining,
    required this.locked,
    this.editableUntil,
  });

  factory Application.fromJson(Map<String, dynamic> json) => Application(
        id: json['id'],
        programId: json['program_id'],
        fullName: json['full_name'],
        birthPlace: json['birth_place'],
        birthDate: json['birth_date'],
        gender: json['gender'],
        address: json['address'],
        phone: json['phone'],
        schoolOrigin: json['school_origin'],
        fatherName: json['father_name'],
        fatherJob: json['father_job'],
        motherName: json['mother_name'],
        motherJob: json['mother_job'],
        parentsIncome: json['parents_income'],
        photoUrl: json['photo_url'],
        status: json['status'],
        editsRemaining: json['edits_remaining'],
        locked: json['locked'],
        editableUntil: json['editable_until'],
      );
}
